### ISSUE TITLE
[HOLD][Expensify/#219254][$2000] Implement periodic background updates on mobile, keeping Onyx in sync

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#4264]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

cc @Julesssss 

## Context
Coming from [this issue](https://github.com/Expensify/Expensify/issues/158830) and also [this one](https://github.com/Expensify/App/issues/4008#issuecomment-886893193).

We have been working to improve the background data trickle of our mobile apps to speed up start-times and improve offline-first behavior. Thusfar, we have been relying on push notifications to wake the backgrounded or killed mobile app and execute some JS to update the app data. However, there are some strict OS limitations on how frequently push notifications can wake an app in the background and perform JS tasks. Furthermore, it seems those limitations are trending towards greater stringency in upcoming OS versions.

So we're hoping to supplement our background app updates through the use of [react-native-background-fetch](https://github.com/transistorsoft/react-native-background-fetch). Early testing shows that the library will allow fetching of report data when killed or backgrounded. We understand that 15 minutes is a best-case scenario and that this will vary widely depending on many variables that are outside of our control.

We need to implement the following: 
- Depend on `react-native-background-fetch` package
- Set up regular (non-headless) task to run in the background periodically (Android | iOS)
- Setup a second headless task which will run when the app is killed (Android)
- For now, both tasks should call the `fetchAllReports` function (later issues should simplify this API call)

## Action Performed:
1. Background the mobile app OR kill the app 
2. Wait 15+ minutes.

## Expected Result:
Our background tasks should run, fetching report data and persisting it to Onyx

## Actual Result:
Nothing 🙃 

## Workaround:
n/a

## Platform:
Where is this issue occurring?

- [ ] Web
- [x] iOS
- [x] Android
- [ ] Desktop App
- [ ] Mobile Web

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by roryabraham** [comment]
@Julesssss Thoughts on making this external?

**Comment by quinthar** [comment]
Love this!

**Comment by mananjadhav** [comment]
If you're planning to make it external, I've got experience with `react-native-background-fetch` and `
`react-native-background-timer`.  I had used it to fetch Ads in the background for a Radio app.

The older version of 'fetch' had issues with the iOS and had to combine 'timer' too.  But updates of 'fetch' won't require a timer now. I would still highlight that these are not 100% reliable for all the different models.

**Comment by Julesssss** [comment]
> Thoughts on making this external?

I think some work can eventually be made external, but for now I would like to keep it internal while I research and test which libraries best meet our requirements, ensuring we implement this critical feature correctly. It doesn't look like `react-native-background-fetch` meets the Android 12 requirement -- so we may need an alternative (other libraries do seem to abide by the OS guidelines).

**Comment by mananjadhav** [comment]
Sure. As far as my experience goes, you won't find a one-size-fits-all lib/solution for all platforms. 

I would recommend creating a wrapper around existing one/multiple libs and keeping it outside of the main App repo.

**Comment by Julesssss** [comment]
I was working on another WIP PR today, will pick this up tomorrow after I submit it.

**Comment by Julesssss** [comment]
Started on this, but switched focus to an API-Error fire and a couple of houry QA issues. 

I have started looking deep into the libraries. One thing unrelated to this issue is that I need to confirm whether the Urban Airship headless service is foreground or not. If not then we should be able to reduce the OS cooldown by making it a foreground headless service (or asking UA to do this) -- will follow up tomorrow.

**Comment by MelvinBot** [comment]
@Julesssss Eep! 4 days overdue now. Issues have feelings too...

**Comment by Julesssss** [comment]
I took a good look at the alternative packages, but `react-native-background-fetch` meets our requirements and has the best documentation/support. For iOS the default native solution is used and for Android It uses JobScheduler under the hood, which the OS will happily let run (as opposed to Services). 

[This alternative](https://github.com/JoaoLSS/react-native-background-worker) `react-native-background-worker` would be even better as it uses the newer `WorkManager` solution, but the package does not support iOS. I don't think it's worth the effort to use one package for each OS selectively, so let's stick with `react-native-background-fetch` for now.

___________________________

Android seems far easier to test as we can manually trigger tasks, so I started there. Some setup is required, which I'll document in this issue for now:
- As expected, without enabling Headless the Background task will only succeed if the app is in the foreground or background. but will cancel immediately if the app is killed (next step will be to enable headless support) 
- Compiling a release build will be necessary to test Headless mode
- Seems like it should be tied to the main App lifecycle, but need to confirm this. I need to make sure we don't initialize the task again when running the headless app
- `TSBackgroundFetch` logs will appear in the native IDE consoles (or Flipper), our logs will appear in the JS console
- We can manually run tasks for testing on Android with `adb shell cmd jobscheduler run -f com.expensify.chat 999`

I also found this excellent source of information on which device manufacturers attempt to further restrict apps and how to get around these issues: https://dontkillmyapp.com/. Finally, [this issue](https://github.com/transistorsoft/react-native-background-fetch/issues/347) has some hints for configuring headless tasks. I'll continue from there tomorrow.  

[WIP PR](https://github.com/Expensify/App/pull/4564/files)

**Comment by Julesssss** [comment]
Had some issues with the example project setup, but I was able to both Headless and non-headless background tasks via the example app:

<img width="974" alt="Screenshot 2021-08-12 at 18 00 28" src="https://user-images.githubusercontent.com/10736861/129237942-9bfbbb4e-e7e3-4f0b-af88-8591196f36e6.png">

<img width="569" alt="Screenshot 2021-08-12 at 18 01 46" src="https://user-images.githubusercontent.com/10736861/129238073-d7c23bce-44f1-48e4-883e-8cfddab6d234.png">

In our app I'm not yet able to get headless tasks to run. Will continue to debug. We report the broadcast intent, but see no other logs.

<img width="740" alt="Screenshot 2021-08-12 at 18 26 25" src="https://user-images.githubusercontent.com/10736861/129241477-69f01961-1523-4889-a8f4-a5d8a139df9a.png">

**Comment by MelvinBot** [comment]
@Julesssss Eep! 4 days overdue now. Issues have feelings too...

**Comment by Julesssss** [comment]
I'm unable to focus on this currently due to allocations, but now that I've settled on a library that suits our needs and works as expected, we can probably export this issue. 

CC @mananjadhav -- perhaps you would be interested in working on this issue next week if I update the task with the latest information and make it external? As this is a relatively complex issue, I think we should double the payment.

**Comment by mananjadhav** [comment]
That's right. @Julesssss Can you share what we have so far in a small comment and the expected goal here? I'll go through it.

**Comment by Julesssss** [comment]
Sure, I'll do this and update the issue but it will need to wait until next week as we're super busy on some high-priority internal projects.

**Comment by MelvinBot** [comment]
@Julesssss Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by Julesssss** [comment]
I've updated the issue with our requirements, following the process to make external.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Julesssss** [comment]
FYI I'm going to take over from the contributor manager that is assigned here

**Comment by Christinadobrzyn** [comment]
Added to Upwork

Internal Upwork job posting - https://www.upwork.com/ab/applicants/1430356207860531200/job-details
External Upwork job posting - https://www.upwork.com/jobs/~01efefc0e091692d09

**Comment by Christinadobrzyn** [comment]
Doubling the price to $500.

**Comment by Julesssss** [comment]
CC @mananjadhav in case you are interested

**Comment by mananjadhav** [comment]
@Julesssss I can pick this starting weekend. Is there a provision to discuss the budget for this task? I feel the effort required might be more than the quote. Let me know if we are open to this.

**Comment by Julesssss** [comment]
That sounds good. Yep, I totally agree this should be valued a bit higher than the average issue and would suggest that we set the price to $1000. I think the library is proven, but I would suggest taking a look at [this PR ](https://github.com/Expensify/App/pull/3877)to work out how to access Onyx in the background. 

It also has steps for building a release version of the app that will let you kill the app entirely (unlike the JS/Chrome builds).

**Comment by mananjadhav** [comment]
That sounds good. 

> That sounds good. Yep, I totally agree this should be valued a bit higher than the average issue and would suggest that we set the price to $1000. Does that sound reasonable?

That sounds good. 

> 
> I think the library is proven, but I would suggest taking a look at [this PR ](https://github.com/Expensify/App/pull/3877)to work out how to access Onyx in the background.
> It also has steps for building a release version of the app that will let you kill the app entirely (unlike the JS/Chrome builds).

Oh, that's great. Thanks for this. I was going through the link and I don't have access to the StackOverflow link mentioned in the PR.

**Comment by Julesssss** [comment]
Oh right, that is one of our internal Stack overflow posts, but you can find similar public posts with pretty much the same info. You just need to search 'How to build iOS/android release' separately. But please let me know if you run into any issues at all here.

Also, when would you like to take on the job? Let me know when you are ready and we will get you hired on UpWork.

**Comment by mananjadhav** [comment]
As I'll start on this by the weekend, I can apply today or Monday. Anything is fine.

I've had an issue earlier running a build on my iPhone which said: 

1. Provisioning Profile is missing and account is not a part of xxxx group
2. Does it need an Apple Developer account to run the release build?

If you can help with these questions that'll be great.

**Comment by Julesssss** [comment]
> Does it need an Apple Developer account to run the release build?

I am pretty sure that's not necessary. See if you can enable the `Automatically manage signing` checkbox here. I think this is possible without being part of the Expensify team...

<img width="1069" alt="Screenshot 2021-09-03 at 11 23 37" src="https://user-images.githubusercontent.com/10736861/131990987-a45a398e-e2a2-41ec-8882-0288b94e071c.png">

**Comment by mananjadhav** [comment]
Thanks for the response. I’ll check tonight and get it working. Android shouldn’t be a problem as I’ve done that earlier.

**Comment by Julesssss** [comment]
@mananjadhav when you get a chance can you apply for the UpWork job too, please.

**Comment by mananjadhav** [comment]
Done

**Comment by mananjadhav** [comment]
@Julesssss This is what I get when I tried Automatically manage signing.

<img width="565" alt="Screenshot 2021-09-03 at 11 21 24 PM" src="https://user-images.githubusercontent.com/3069065/132047279-e9126495-26ae-4738-9b2e-596303ee4341.png">
<img width="625" alt="Screenshot 2021-09-03 at 11 21 15 PM" src="https://user-images.githubusercontent.com/3069065/132047295-a7b1a18e-45e2-47ff-852c-bd1f186f71d4.png">

**Comment by Julesssss** [comment]
Hi @mananjadhav, I have asked in [Slack here](https://expensify.slack.com/archives/C01GTK53T8Q/p1630919241473800). 

Thankfully iOS is the simpler platform to implement, so I am happy to manually test iOS for you if necessary.

**Comment by mananjadhav** [comment]
Okay, I struggled with the release setup of iOS over the weekend with some other issues. I'll be starting back to it by tomorrow. Thanks 👍

**Comment by Christinadobrzyn** [comment]
Hired @mananjadhav in Upwork for this! 🎉

**Comment by mananjadhav** [comment]
I ran their demo example and it looks okay. I'll now start with Expensify and close it by the weekend. Should have something to test by then with Android.

**Comment by Christinadobrzyn** [comment]
Hi @mananjadhav do you have an update on this for the team?

**Comment by mananjadhav** [comment]
Still WIP. Should have by Wednesday.

**Comment by mananjadhav** [comment]
@Julesssss Can you help with the release version of Android? Debug version works fine, but the release doesn't. 

I am aware of the general key signing process and updating in `build.gradle` but I am not sure if Expensify follows something specific. I am unable to access the SO link mentioned in the push notification task.

**Comment by Julesssss** [comment]
Sure. There is a bit of [info here](https://github.com/Expensify/App/blob/main/PERFORMANCE.md#how-to-create-a-release-build-on-android). We're not doing anything special here though.

Can you share more info on the gradle error?

**Comment by mananjadhav** [comment]
Okay, I'll try this info and get back. This looks similar to what I had on Google. What I wasn't sure was if I could replace the values in `build.gradle` with hardcoded values.

**Comment by Julesssss** [comment]
Yeah, you can hardcode the key info in `build.gradle` and just clear it before pushing the PR for review.

**Comment by kadiealexander** [comment]
Hi @mananjadhav, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

**Comment by Christinadobrzyn** [comment]
Not overdue - on hold

**Comment by mananjadhav** [comment]
I wasn’t available last week due to some exigency. Quick update, release build worked. Task doesn’t get auto triggered in the killed state but if I simulate the trigger it works fine. I’ll raise a draft PR today.

**Comment by Julesssss** [comment]
No worries @mananjadhav, thanks for the update! Please let me know if I can help in any way with testing, etc.

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by Christinadobrzyn** [comment]
On hold for n6

**Comment by mananjadhav** [comment]
I wasn’t available last week due to health issues. But yeah testing as of now. 

Resumed now. I’ll raise the PR by tomorrow.

**Comment by Christinadobrzyn** [comment]
N6 hold is lifted - this job will include a $250 bonus for the n6 hold. 

Still waiting on a PR from @mananjadhav

**Comment by Christinadobrzyn** [comment]
just checking in on this @Julesssss and @mananjadhav - it looks like we're still working through the [PR](https://github.com/Expensify/App/pull/5953)?

**Comment by Julesssss** [comment]
@Christinadobrzyn yep, I've caused the delay here due to a busy period at work, then some OOO due to an injury. I'll be back onto this soon.

**Comment by Christinadobrzyn** [comment]
Sounds good @Julesssss! Looks like melvin is looking for an update - sounds like we're still reviewing the PR.

**Comment by Julesssss** [comment]
Yep, unfortunately, I haven't had time to review this due to being out for that injury, then the IPO week followed by OOO time. I'm back from the 29th. Sorry for the delay @mananjadhav!

**Comment by kidroca** [comment]
This ticket is related to some of my work so I'll do whatever I can to help move it forward

I took a look at the PR here: #5953 and left a review
So far it looks on the right track but it just proves the concept of the idea - the background task that is defined would make a sample request and then log to the console
It should instead do something similar to the stuff we do on startup - make a call to fetch reports and then persist data to Onyx
Probably the task need to just import `Report` and call the appropriate method

Since the documentation states that if we receive a `timeout` event from `BackgroundFetch` we should abort immediately, we might need to implement something similar to what's proposed here: https://github.com/Expensify/App/pull/6388#issuecomment-991236205

**Comment by mananjadhav** [comment]
@kidroca Thanks for the comments. Yes, the PR is still in draft and requires cleanup. 

> It should instead do something similar to the stuff we do on startup - make a call to fetch reports and then persist data to Onyx

We wanted to start a basic function call and then convert

**Comment by Christinadobrzyn** [comment]
Looks like the [PR](https://github.com/Expensify/App/pull/5953) is still under review.

**Comment by Julesssss** [comment]
No change

**Comment by Julesssss** [comment]
Hi @mananjadhav, maybe we should think about closing this for now?

**Comment by roryabraham** [comment]
> maybe we should think about closing this for now?

Why?

**Comment by mananjadhav** [comment]
Was a bit unwell, will push an update by Fri.

**Comment by Julesssss** [comment]
Hey @mananjadhav, how is this looking?

**Comment by mananjadhav** [comment]
Pushing the update today.

**Comment by Julesssss** [comment]
I spoke with @mananjadhav about dropping this issue and he agreed that it was for the best. Thanks for your time looking into this Manan, we will pay out $500 for the time you spent (once the issue is solved).

**Comment by Julesssss** [comment]
This issue is more complex than I initially realized, so I would also like to raise the bounty before seeking another contributor.

**Comment by kidroca** [comment]
I know the current task is regarding fetching report data in the background, but once that is setup we can easily use it to push persisted messages when the user regains connectivity (instead of waiting for them to open the app again): 
- https://github.com/Expensify/App/issues/6690

### Proposal 
Extract a lib to configure and register a background-fetch task 
Use existing logic to sync any report updates. Pretty much same as the reconnect callback here:
https://github.com/Expensify/App/blob/75c7986c1b3fcb1683c42adc2a0079ddc58e0b2d/src/libs/actions/Report.js#L1397-L1398
Use `requiredNetworkType: BackgroundFetch.NETWORK_TYPE_ANY` to skip running the task if the phone is offline

The implementation would try to do roughly this:

```js
BackgroundFetch.configure({
  minimumFetchInterval: 15
}, (taskId) => {  
  Log.info(`[Background] running background task: ${taskId}`);
  fetchAllReports().finally(() => BackgroundFetch.finish(taskId))
}, async (taskId) => {  // <-- Event timeout callback
  // This task has exceeded its allowed running-time.
  // You must stop what you're doing and immediately .finish(taskId)
  Log.info(`[Background] task timed out: ${taskId}`);
  BackgroundFetch.finish(taskId);
});
```

Furthermore...
When the iOS app is launched by a background task it would mount the root component and execute React lifecycle events
Mounting the React tree while in the background is something we'd like to avoid (as explained in [this slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1645708049935549))
RN firebase has documented the problem and a possible solution here: https://rnfirebase.io/messaging/usage#background-application-state
We can apply their suggestion and alter `AppDelegate.m` so it passes an `isHeadless` prop to the root component
The root (`<Expensify />`) would use it to skip mounting the rest of the App tree

**Comment by Julesssss** [comment]
This is great, thanks for explaining the proposal. 

So to clarify, this issue would now be to **A)** Implement periodic background updates ensuring that iOS is not mounting the app when running the background task, and **B)**  Pushing cached messages in the background instead of awaiting the next App open. I wasn't sure that 2 will be possible to implement yet though, does that sound reasonable to you @kidroca?

We should also increase the bounty to $2000 given the additional tasks, and the complexity of implementing the above.

**Comment by kidroca** [comment]
We can have B) as a continuation work, definitely a separate PR 
I'm just bringing it back up because it would be a lot easier to implement if we already have A) and I see value in it

**Comment by Julesssss** [comment]
Gotcha. That's fine. 

If you would like to continue with this issue I'll go ahead and assign you.

**Comment by MelvinBot** [comment]
📣 @kidroca You have been assigned to this job by @Julesssss!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by Julesssss** [comment]
Hi @Christinadobrzyn, would you please hire kidroca on UpWork and update the amount if that's possible? Thanks!

**Comment by kidroca** [comment]
I can't find an Upwork job about this ATM, can you post it here when it's available

**Comment by Julesssss** [comment]
I think it is [this one](https://www.upwork.com/jobs/~01efefc0e091692d09)

**Comment by kidroca** [comment]
Thanks, I'll wait for the amount to be updated before applying

**Comment by kidroca** [comment]
Seems there are some discussion around the topic of this issue: https://expensify.slack.com/archives/C01GTK53T8Q/p1646082982919629

@Julesssss
New idea: Instead of using `background-fetch` which tries to sync periodically (poll) we can receive a data-only notification

Notifications payloads are small so we can't use them to update Onyx directly but we can use them to wake-up App and make it execute specific GET call and sync itself depending on the notification payload

Now that I think of it we don't even need to use `data-only` notifications. Currently we can implement the same when a regular notification wakes up App.
When a notification is received run `fetchAllReports` - ⚠️but only if reports haven't been fetched in the last 30 min. So receiving multiple pushes would not trigger too much fetching
At the end of `fetchAllReports` we capture a timestamp that will serve for this check

Some cons for background-fetch:
- it's based on polling - increased battery usage (though I'm not sure how much)
- not strict - you can set an interval like 30 minutes. but events can happen a lot less often or regularly
- if the user kills the app iOS would no longer fire events and wake the app up to let us try to sync 

Compared to Notifications:
- not based on polying - a notification would wake up the app anyway and cost batter so we might as well try to fetch some data
- easier to test - if we trigger `fetchAllReports` when a notification is received - it's visible in both server and local logs
- if the app was killed a notification would wake it up and give it the chance to sync

Ideally at some point this can be switched to `data-only` notifications so we only run fetch when it's guaranteed there's data we don't have locally

---

I'm OK for implementing `background-fetch` the fetching logic would be reusable no matter what strategy triggers it. And things like not mounting the whole React tree are still needed

**Comment by roryabraham** [comment]
> New idea: Instead of using background-fetch which tries to sync periodically (poll) we can receive a data-only notification

I can provide some additional context from my understanding of push notifications here @kidroca. There's a lot of inconsistent information out there, but this is my understanding.

The flag that makes a difference on iOS is called `content_available`, and that informs that iOS app that it needs to wake from the background and perform some operations. That flag can be passed with "silent" data-only notifications (i.e: no `alert`), or with "loud" notifications that will be displayed by the OS and included in the messaging center. In both cases, the flag informs the OS that it needs to perform a background update.

You may find some outdated information that says `content_available` only works when an app is backgrounded but not completely closed, but IIRC that changed as of iOS 10 or 11, and it works when the app is completely killed. We send the `content_available` flag with all our report comment push notifications, so we already have the background data updating implemented via push notifications. _However_, from the [Apple Documentation](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app?preferredLanguage=occ):

> The system treats background notifications as low priority: you can use them to refresh your app’s content, but the system doesn’t guarantee their delivery. In addition, the system may throttle the delivery of background notifications if the total number becomes excessive. The number of background notifications allowed by the system depends on current conditions, but don’t try to send more than two or three per hour.

There are also other variables that come into play. For example, from my testing background notifications do not work on iOS at all if the phone is in low power mode. I've also found that by saying "the system doesn't guarantee their delivery", Apple really means "the OS doesn't guarantee that your app will actually be awoken in the background when the notification is received".

---

So the problem in this issue is that background data updates on iOS are just not consistent or reliable. This issue is really meant to bootstrap the background updates we already have through a more reliable, OS-sanctioned channel via `background-fetch`.

**Comment by roryabraham** [comment]
A bit more detail too – when we receive a push notification, it includes the reportID and the reportAction. We just update the report with the single reportAction included in the notification payload. We could improve that and fetch all the new actions for the report, or _maybe_ all actions for all reports. But I think the OS is pretty stringent in the amount of time you have to complete callbacks. It even might forbid the app from doing network requests in the background update, we would have to test and be sure that's allowed. If we can, though, this update strategy would make sense to me:

|type|update|
|---|---|
|push notification|All actions for the single report with a new chat|
|periodic (background-fetch)|All actions for all reports|

**Comment by Julesssss** [comment]
I agree this is a fair alternative, but also that notifications can be unreliable. For any reason the OS can throttle them and background updates help to handle these cases. 

We do have other issues for the silent notification, but I think it's out of scope for this issue. Once this is done we can look at improving the notification logic in addition.

**Comment by kidroca** [comment]
> There are also other variables that come into play. For example, from my testing background notifications do not work on iOS at all if the phone is in low power mode

Power mode affects all background activities and [background fetching/refreshing is disabled as well](https://www.makeuseof.com/iphone-low-power-mode/#low-power-mode-disables-background-app-refresh) in low power mode

> Apple really means "the OS doesn't guarantee that your app will actually be awoken in the background when the notification is received".

What's guaranteed though is the background push would be processed [either in the background or when App is launched](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app#3526627)

> So the problem in this issue is that background data updates on iOS are just not consistent or reliable.

Background **fetch** is actually subject to more rate limiting than background pushes - specifically app usage
Background **push** is subject to the same throttling as fetch, but is not gated by app usage
What does that mean:
Background fetch might run as little as 0 or 1 times per day, on the other hand receiving a background push might be delayed somewhat so that receiving 2-3 pushes in 30min. would only launch the app after the last push

I find **push** more reliable 

> This issue is really meant to bootstrap the background updates we already have through a more reliable, OS-sanctioned channel via `background-fetch`.

When App launches we still won't know whether Report data is fresh, and we'd still have to run `fetchAllReports` defeating the purpose of any prior polling that background fetch might have performed
If we're still going to trigger `fetchAllReports` when App is foregrounded, it makes little sense to poll the same action in the background

> We could improve that and fetch all the new actions for the report, or _maybe_ all actions for all reports.

I don't think we need to fetch actions - they are already getting delivered through Push Notifications. Pushes have TTL even if we've missed some (being offline) they would arrive. But even if that fails we have `fetchAllReports` when App is foregrounded 

We can use background push to trickle and react to other updates silently - Reports last read, max actions, etc...

> But I think the OS is pretty stringent in the amount of time you have to complete callbacks

Both fetch and push have a 30 sec. limit

> It even might forbid the app from doing network requests in the background update, we would have to test and be sure that's allowed

You are allowed to use a notification to let an app know there's new data it can fetch from the server and fetch it in the provided 30sec. wake window. Bonus points if you're done earlier and signal the system it's free to re-suspend the app

> periodic (background-fetch)	All actions for all reports

I don't know about you but I see a lot of wasted bandwidth here 
Unless we have a sure way to sync only stuff that we're lacking
If we're going to fetch all reports each time the app is foregrounded background-fetch loses even more value 

> I agree this is a fair alternative, but also that notifications can be unreliable. For any reason the OS can throttle them and background updates help to handle these cases.

background updates are arguably throttled even more
The aim of the OS is to predict and run a background updated shortly before you foreground the app (based on user habits), but that's not guaranteed to happen every time or even 50% of the time 
While a push notification might be delayed, but will be processed at the time the app is being foregrounded

---

Background fetch tries to runs tasks periodically so that App can be fresh just before the user foregrounds it
It's great for stuff like fetching news, stock prices, social media feed - a constant stream of information where there's always something new to be retrieved

The world of a messaging / mailing it's not following the same dynamics. Modern apps use efficient budgeting and fetch new content when it becomes available, they are not blindly polling for new data as often as possible

---

[Actual apple suggestion](https://developer.apple.com/videos/play/wwdc2019/707/?time=820)
> Use background push as best effort way to download content
Can always download when app re-enters foreground

**Comment by roryabraham** [comment]
Okay, so given all that @kidroca, what's your suggestion?

**Comment by kidroca** [comment]
One you login on mobile all relevant Report actions are immediately fetched 
As explained on [this slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1646085456104599?thread_ts=1646082860.227219&cid=C01GTK53T8Q) any new report actions are delivered through Pusher and/or push notifications

Most of the time when App is focused we'll be up to date on actions
Since actions are already pushed to us I don't think we need to periodically fetch all report actions
Instead when the app is foregrounded it should be able to quickly fetch whatever might be missing - most of the time this would be nothing, so a `fetchAllReports` that returns 10 reports and 500 actions that you already have is not ideal 

The idea I like so far is: Whenever a push notifications carrying a report actions is received - it re-fetches Report information regarding its own Report. The updated information would allows to make decision on whether we need to fetch anything more for the given report (if it's a Report we somehow don't have locally fetch it's latest 50 actions)

If we can somehow make `fetchAllReports` work with a timestamp so that we can call it to get everything new since the morning (when we last called it) it would be great and we can use that with background fetch as well as when App is foregrounded

If we can't make `fetchAllReports` get only updates, then I would leave it the way it is. Rely on having up to date information from pushes and trigger `fetchAllReports` when App is foregrounded 
- if we're already up to date, we're good 
- if we aren't we're getting there shortly

**Comment by kidroca** [comment]
A single call `fetchAllReports` is enough to get you up to date and then you can carry on receiving further updates from pushes

**Comment by roryabraham** [comment]
> If we can somehow make fetchAllReports work with a timestamp so that we can call it to get everything new since the [last time we called it]

This should be possible, sounds reasonable to me.

> Whenever a push notifications carrying a report actions is received - it re-fetches Report information regarding its own Report

There used to be an issue for this: https://github.com/Expensify/App/issues/4008 and it got closed in preference of this one.

So it sounds like @kidroca is suggesting we should:

- Close this issue and reopen https://github.com/Expensify/App/issues/4008. Having forgotten the context for why https://github.com/Expensify/App/issues/4008 was closed, it seems obvious to me that we should make the most of our inconsistent background updates, and that only adding a single `reportAction` to a report is not a great use of that opportunity.
- Optimize `fetchAllReports` to take a timestamp and only provide `reportActions` added to any of your reports since a given time. It's unclear to me whether this will lead to a meaningful performance improvement, but it's possible. It would also depend on how that change affects the speed of the query. In any event, I think this should really be a separate issue.

**Comment by Julesssss** [comment]
Okay, thanks for taking the time to explain all that. I still think there is some value to both, but I'm on board with the above ^ 

Let's move this bounty to the #4008 and reassess once we have push notification backup 👍 (I think the bounty still makes sense as it resolves our main issue)

**Comment by roryabraham** [comment]
Another thought:

> What's guaranteed though is the background push would be processed [either in the background or when App is launched](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app#3526627)

Let's say 5 different reports received background push updates, but due to OS throttling, none of them ran while the app was backgrounded/closed. When the app comes back to the foreground, we would then do 5 network requests to update each of the 5 reports we know are out-of-date. It would be preferable to have a single `updateReports` API that could take a list of `reportID`s and update them all with a single request.

**Comment by roryabraham** [comment]
And yet another scenario:

1. We receive 5 push notifications for a report when we're at the airport and the app is closed in our pocket.
1. We go on the plane, and put the phone in airplane mode.
1. We open the app on the plane to see the 5 messages we missed, but since the network request to update that report will fail, we don't see them ❌ 

So if we're offline, we'll need to fallback onto the push notification payloads so that we don't get "ghost messages" that's visible via a push notification but not in the app. Or better yet, _always_ add the reportAction to Onyx right away, then attempt a network request to update all the actions on that report.

**Comment by kidroca** [comment]
Thanks for bearing with me

Let's agree the goal of background updates is to have as up to date data as possible when App is brought back to focus

A summary why pushes are satisfying the above goal:
- they work for logged in users - being logged in means you've been through the "big" data fetching (that happens after login)
- new messages are typically being pushed to you - Pusher or Push Notifications

Apple calls this "best effort to be fresh"
There are circumstances, as pointed out, where neither bg fetch or push would work making a foreground logic necessary - to resync at least once when App is relaunched/foregrounded

When we talk about Reports, there are 2 kinds of report data - actions and the report metadata itself

@roryabraham 
When a new comment is received via push notification (alert or not), we definitely would keep the payload and merge it to Onyx. I supposed it already works that way. In addition to that we can try to fetch Report metadata (not actions) - or something else that makes sense:
- if we're offline we can write the request to the persisted queue (now this is something that you can later flush using background fetch, or when app is foregrounded, whichever is first). A more simple solution might be to take a note and do a "big" refresh when App is foregrounded
- does it make sense to fetch report metadata? - max action sequence number can be derived from the message, but what about anything else. If it doesn't make sense to fetch metadata we can use the time to do something else (or nothing)
- if for some reason we don't have information for the report, e.g. someone starts a chat with us, we would fetch the info. Maybe it would be a good idea to include more metadata in the first report action to be able to construct a report out of it (in case we're offline and can't fetch)
- if for some reason we don't have a report and it's not the first action, then if possible we should fetch metadata and latest actions

> When the app comes back to the foreground, we would then do 5 network requests to update each of the 5 reports we know are out-of-date.

Depending on what we're fetching it's probably not a big deal, and the requests would happen in parallel

**Comment by kidroca** [comment]
One thing that we have to make sure if not already so is to prevent notification collapsing.
Notification collapsing (apns-collapse-id) is when a newer notification replaces an older one
Data only notifications (no alert, no sound, content-available: 1) are non collapsible by default, but regular notifications are.
Resuming connectivity would receive only the latest message of a collapsible notification
What makes sense for a chat App is to stack notifications in groups using `thread-id` - notifications for the same report can be visually stacked together. Not sure about what's the Android counterpart for this

**Comment by roryabraham** [comment]
> When a new comment is received via push notification (alert or not), we definitely would keep the payload and merge it to Onyx. I supposed it already works that way. In addition to that we can try to fetch Report metadata (not actions)

Hmmm, why not actions (for that specific report)?

> if we're offline we can write the request to the persisted queue

I like this idea!

> [The persisted queue] is something that you can later flush using background fetch

Love this idea! This would be great if you had a spotty network. Let's say a network request fails while a push notification is processing – we would add that request to the persisted queue and then, even if you didn't open the app, the persisted request queue would attempt to flush every 15-30 minutes with background fetch.

> does it make sense to fetch report metadata? - max action sequence number can be derived from the message, but what about anything else

Maybe participants! We could maybe also pass up-to-date report metadata along with the push notification

**Comment by roryabraham** [comment]
> One thing that we have to make sure if not already so is to prevent notification collapsing.

This is definitely something that is not already handled. I'll create an issue to follow-up.

**Comment by kidroca** [comment]
> Hmmm, why not actions (for that specific report)?

We presume we already have all the actions predating the one from the notification
We can have some sanity check and if we don't then yeah fetch previous actions and meta

Another topic to consider is the redesign of the **read** queue: https://expensify.slack.com/archives/C01GTK53T8Q/p1645658235688649

1. App receives updates for almost everything through pusher 
2. When App is re-focused, currently mounted components would know what data they can try to refresh and trigger a fetch. _(Typically on a Report page - everything related to the current reportID)_
3. Same as 2) when we transition from offline to online
4. As a consequence this simplifies the read queue and makes it unnecessary to persist any read requests. When we're offline we don't make read requests, when we're back online we only fetch data for the latest page we're on

Point 4) is conflicting with proposals here suggesting to persist a read request, though it's a situation that might have been overlooked and deserves to be an exception

We decided in favor of 4) because the most important thing is to get the current screen you're coming back to be up to date or be quickly made that way, which 2) and 3) achieve. Then doing a general `fetchAll` would take care of getting any actions we might have missed being offline

**Comment by Julesssss** [comment]
> Resuming connectivity would receive only the latest message of a collapsible notification
What makes sense for a chat App is to stack notifications in groups using thread-id - notifications for the same report can be visually stacked together. Not sure about what's the Android counterpart for this

Similar behavior exists on Android -- but we have to configure it similar to how you said it works for iOS: We can 'update' notifications by making a push with the same uniqueID. This is how all other chat apps work usually on Android, but we haven't implemented this yet. We could either send the entire 'thread' of unread message data with each push, or instead build this unread message cache locally.

**Comment by Julesssss** [comment]
> Point 4) is conflicting with proposals here suggesting to persist a read request, though it's a situation that might have been overlooked and deserves to be an exception

Would you say we're currently blocked by the read network queue proposal in that case? Or do we know enough about the eventual solution to tailor this issue?

**Comment by kidroca** [comment]
> Would you say we're currently blocked by the read network queue proposal in that case? Or do we know enough about the eventual solution to tailor this issue?

I don't think we're blocked 
The idea to flush persisted requests in a background task can be tackled separately
We can start with the outline here: https://github.com/Expensify/App/issues/4264#issuecomment-1055888496 

1. When a new comment is received via push notification (alert or not), we merge it to Onyx 
2. Check whether we need to fetch anything more about the involved Report
3. Or use the time to do arbitrary work like `fetchAllReports` if we haven't done so recently
4. Or do nothing - scoring nice app points from the OS
5. Inform the OS we're done doing background work.

If we're happen to be offline by the time the background execution run we can just do nothing for the moment

⚠️ Regarding 5) - Urban Airship seems to lack this capability 

See how other libraries have a hook to inform the OS they are done doing background work:

https://github.com/zo0r/react-native-push-notification#usage
```js
    // (required) Called when a remote is received or opened, or local notification is opened
    notification.finish(PushNotificationIOS.FetchResult.NoData);
```

https://wix.github.io/react-native-notifications/api/general-events#registernotificationreceivedbackground
```js
Notifications.events().registerNotificationReceivedBackground((notification: Notification, completion: (response: NotificationCompletion) => void) => {
  console.log(JSON.stringify(notification.payload));

  // Calling completion on iOS with `alert: true` will present the native iOS inApp notification.
  completion({alert: true, sound: true, badge: false});
});
```
> Should call completion function on iOS

https://notifee.app/react-native/docs/events#background-events
`notifee` uses Promises to track this, when a promise is over the OS is informed we're done: 
> The [onBackgroundEvent](https://notifee.app/react-native/reference/onbackgroundevent) method is used to register a callback handler which will be executed whenever a background event is sent. Executing the callback handler in the background requires device resources so it is important that any code is executed efficiently & quickly.

Unless Urban Airship allows finishing background work, we'll have to swap it for another lib
We can still use it as a service, but handle notifications (or at least their payload) with another lib

**Comment by kidroca** [comment]
> We have a pretty good relationship with UA, we can ask for that functionality or even submit a PR to their library if necessary.

I find UA's Push Notifications API for react-native lacking other useful functionalities 
- when app is foregrounded get all unprocessed notifications (like when you open App from it's icon and not from a notifications in the notification bar)
- clearing notifications - we'd probably want to automatically dismiss a notification when user sees a chat for which they have a notification in the notification bar
- modify notification content - we might never needs this
- create local notifications - we might have the need to raise a notification from the background - like "We failed to sync some information in the background because you seem to be Offline, tap here when you're back online so we can try again". Or when the user leaves App while messages are sending and OS asks us to stop - "We couldn't deliver you're last 3 message, tap here to retry"
- finishing background work - already explained [here](https://github.com/Expensify/App/issues/4264#issuecomment-1057112988)

I guess this ticket is blocked if we need to have this on the UA react-native API
At the same time we can work with any Push Notifications library by modifying the code here: https://github.com/Expensify/App/blob/0e3bbe77a3ed497b49ffb3ee32d952e480bda448/src/libs/Notification/PushNotification/index.native.js#L16-L46
We can avoid efforts writing PRs, but just submit feature requests, when some day features are implemented we can switch the above code to use UA
I don't know about you but I don't feel confident writing native code for both Android and iOS, or at least I don't think I'll create something better than what already exists

**Comment by kidroca** [comment]
Hmm actually it was just hard finding these docs: https://docs.airship.com/reference/libraries/react-native/latest/urbanairship-react-native/classes/_urbanairship_.urbanairship.html

There are methods covering above items like:
- https://docs.airship.com/reference/libraries/react-native/latest/urbanairship-react-native/classes/_urbanairship_.urbanairship.html#getactivenotifications
- https://docs.airship.com/reference/libraries/react-native/latest/urbanairship-react-native/classes/_urbanairship_.urbanairship.html#clearnotification

Still it doesn't seem possible to notify OS when we're done doing work in the background

**Comment by Christinadobrzyn** [comment]
(I want to remove the Overdue label)  - @Julesssss or @kidroca can you provide an update on the next steps for this job? thanks!

**Comment by kidroca** [comment]
I think the issue is now blocked by the planned networking changes and receiving data through Pusher

We then have to decide how background updates work
- API sends background push -> device tries to resync
- or device does some resync work each time a notification arrives

And decide how the update logic works
- make background sync an exception case and use HTTP responses in order to complete a background task as fast as possible
- when App wakes up in the background still make a Pusher connection and wait (30 sec. max) to receive messages - always consuming the max allowed wait time 

_It seems we're moving in a direction where the 1 bullet won't be an option, because API responses would just not carry the information_

**Comment by Julesssss** [comment]
Still in planning

**Comment by Julesssss** [comment]
No changes

**Comment by Julesssss** [comment]
Same as above

**Comment by Julesssss** [comment]
Still held for now.

**Comment by Julesssss** [comment]
Held on notification project

**Comment by trjExpensify** [comment]
Can we pop a `[Hold #219254]` in the title to link to the notifications project, so it's clear what this is held on?

**Comment by Julesssss** [comment]
Yeah, I'll do this for all these issues.

**Comment by Christinadobrzyn** [comment]
Hey @Julesssss @kidroca would you be able to give an update on this? I'm not really sure what we're doing with this job at the moment. Thanks!

**Comment by kidroca** [comment]
I'm not sure as well, but I think we should change the ticket description or open a new one, since things changed significantly

AFAIK we don't pursue **periodic** attempts to sync the app, but instead trickle down updates via Pusher or Push notifications

**Comment by Julesssss** [comment]
Hey @Christinadobrzyn this issue should be on hold for the moment, so I just updated the title

**Comment by Julesssss** [comment]
Still held for now

**Comment by Julesssss** [comment]
Will still be held for a while. It might be covered by the notifications doc #2, but more likely it will follow it instead.

**Comment by Julesssss** [comment]
Not overdue. it's on hold

**Comment by Julesssss** [comment]
Getting closer to unblocking this.

**Comment by Julesssss** [comment]
Making progress with other notification issues, and this one remains one of the lower priorities.

**Comment by Christinadobrzyn** [comment]
@Julesssss would you have an update on this one?

**Comment by Julesssss** [comment]
This is now completely on hold as we might have an alternate solution to the [missing onyx data](https://github.com/Expensify/Expensify/issues/255265#issuecomment-1564597571) problem.

**Comment by arosiclair** [comment]
Closing this out in favor of #15058. We considered this background-fetch approach, but there's a lot of concern about increased server load so we won't be using it. More context [here](https://expensify.slack.com/archives/C04E99DVBTP/p1685021898689099?thread_ts=1684585843.702709&cid=C04E99DVBTP).

**Comment by mananjadhav** [comment]
@arosiclair @Julesssss I had spent time on this issue earlier, also worked on a PR. We had decided to for a payout [here](https://github.com/Expensify/App/issues/4264#issuecomment-1049983171), when the issue is resolved. Is this still valid @Julesssss ?

cc - @Christinadobrzyn

**Comment by Julesssss** [comment]
Yep, that still seems valid. @Christinadobrzyn would you mind handling that payment please?

**Comment by Christinadobrzyn** [comment]
Here's a new job posting for this: 

Internal: https://www.upwork.com/ab/applicants/1669723658502512640/job-details
External: https://www.upwork.com/jobs/~013d0af18888aa9942

@mananjadhav I invited you to this one - can you please accept and I'll pay out? Thanks!

**Comment by mananjadhav** [comment]
Thanks @Christinadobrzyn. Accepted

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks! I paid you in Upwork.

**Comment by Christinadobrzyn** [comment]
Hey @Julesssss should we close this?

**Comment by Julesssss** [comment]
Yeah 👍
